bob/src/hw_metrics_collector.rs

```diff
@@ -179,53 +183,165 @@ enum CommandError {
+                                err = Some(msg);
+                                break;
+                            }
                         }
                     }
+                    if err.is_none() {
+                        return Ok(f_iowait * 100. / sum);
+                    }
+                } else {
+                    let msg = format!("CPU stat format in {} changed", CPU_STAT_FILE);
+                    err = Some(msg);
                 }
-            }
+            },
             Err(e) => {
-                self.iostat_avl = false;
-                return Err(CommandError::Primary(e));
+                err = Some(e);
+            }
+        }
+        self.procfs_avl = false;
+        Err(CommandError::Primary(err.unwrap()))
+    }
+}
+
+const DIFFCONTAINER_THRESHOLD: Duration = Duration::from_millis(500);
+
+struct TimeContainer {
+    last: Option<Instant>,
+    prelast: Option<Instant>,
+}
+
+enum TimeMoment {
+    Last(Duration),
+    Prelast(Duration),
+}
+
+impl TimeMoment {
+    fn elapsed_ms(&self) -> f64 {
+        match self {
+            TimeMoment::Last(dur) => dur.as_millis() as f64,
+            TimeMoment::Prelast(dur) => dur.as_millis() as f64,
+        }
+    }
+}
+
+impl TimeContainer {
+    fn new() -> Self {
+        Self {
+            last: None,
+            prelast: None,
+        }
+    }
+
+    fn update(&mut self, now: Instant) {
+        self.prelast = self.last;
+        self.last = Some(now);
+    }
+
+    fn tick(&mut self) -> Option<TimeMoment> {
+        let now = Instant::now();
+        if let Some(last) = self.last {
+            let elapsed = now.duration_since(last);
+            if elapsed > DIFFCONTAINER_THRESHOLD {
+                self.update(now);
+                return Some(TimeMoment::Last(elapsed));
+            } else if let Some(prelast) = self.prelast {
+                let elapsed = now.duration_since(prelast);
+                self.update(now);
+                return Some(TimeMoment::Prelast(elapsed));
             }
         }
-        self.iostat_avl = false;
-        Err(CommandError::Primary("Failed to get iowait from iostat".to_string()))
+        self.update(now);
+        None
+    }
+}
+struct DiffContainer<T> {
+    last: T,
+    prelast: T,
+}
+
+impl<T> DiffContainer<T>
+where T: Default + 
+         std::ops::Sub<Output = T> +
+         Copy,
+{
+    fn new() -> Self {
+        Self {
+            last: T::default(),
+            prelast: T::default(),
+        }
+    }
+
+    fn diff(&mut self, new: T, elapsed: &TimeMoment) -> T {
+        match elapsed {
+            TimeMoment::Last(_) => {
+                let diff = new - self.last;
+                self.prelast = self.last;
+                self.last = new;
+                diff
+            },
+            TimeMoment::Prelast(_) => {
+                let diff = new - self.prelast;
+                self.prelast = self.last;
```

Comment:
We don't need to update values if the diff calculated from prelast value. This logic is tightly coupled with the update logic in `TimeContainer`, so I suggest to combine them into a single struct